File: experiment-5/runs/language=clojure_model=claude-opus-4-8_tooling=none/rep2/test/brazilian_soccer/queries_test.clj
;; =============================================================================
;; brazilian-soccer.queries-test
;; -----------------------------------------------------------------------------
;; CONTEXT
;;   BDD (Given/When/Then) tests for the query/aggregation layer of the
;;   Brazilian Soccer MCP server. Uses the hand-verifiable fixture dataset in
;;   brazilian-soccer.fixtures so every expected count is exact, plus a few
;;   smoke checks against the real datasets.
;; =============================================================================
(ns brazilian-soccer.queries-test
  (:require [clojure.test :refer [deftest testing is]]
            [brazilian-soccer.queries :as q]
            [brazilian-soccer.data :as data]
            [brazilian-soccer.fixtures :as fx]))

(def ms fx/matches)
(def ps fx/players)

(deftest search-matches-by-team
  (testing "Given matches, When searching by team, Then all their games across competitions return"
    (is (= 7 (count (q/search-matches ms {:team "Flamengo"})))) ; 5 Brasileirão + Copa + Libertadores
    (is (= 4 (count (q/search-matches ms {:team "Santos"})))))

  (testing "When filtering by season, Then only that season returns"
    (is (= 6 (count (q/search-matches ms {:competition "Brasileirão" :season 2019})))))

  (testing "When filtering home only, Then away games are excluded"
    (is (= 2 (count (q/search-matches ms {:home "Flamengo" :season 2019
                                          :competition "Brasileirão"})))))

  (testing "When filtering by date range, Then only matches in range return"
    ;; May 2019: Flamengo 3-0 Palmeiras (1st) and the Copa do Brasil final (15th)
    (is (= 2 (count (q/search-matches ms {:from "2019-05-01" :to "2019-05-31"})))))

  (testing "Then results are sorted most-recent first"
    (let [dates (map :date (q/search-matches ms {:team "Flamengo"}))]
      (is (= dates (reverse (sort dates)))))))

(deftest head-to-head-record
  (testing "Given two rivals, When computing H2H, Then record is exact"
    (let [{:keys [record]} (q/head-to-head ms "Flamengo" "Palmeiras")]
      (is (= 3 (:total record)))
      (is (= 1 (:a-wins record)))
      (is (= 1 (:b-wins record)))
      (is (= 1 (:draws record)))
      (is (= 4 (:a-goals record)))
      (is (= 3 (:b-goals record))))))

(deftest team-statistics
  (testing "Given a team/season, When computing stats, Then W/D/L and goals match"
    (let [s (q/team-stats ms "Flamengo" {:season 2019 :competition "Brasileirão"})]
      (is (= 4 (:matches s)))
      (is (= 3 (:wins s)))
      (is (= 1 (:draws s)))
      (is (= 0 (:losses s)))
      (is (= 11 (:goals-for s)))
      (is (= 1 (:goals-against s)))
      (is (= 10 (:points s)))
      (is (= 75.0 (:win-rate s)))))

  (testing "When venue is home, Then only home games are counted"
    (let [s (q/team-stats ms "Flamengo" {:season 2019 :competition "Brasileirão"
                                         :venue :home})]
      (is (= 2 (:matches s)))
      (is (= 2 (:wins s)))
      (is (= 8 (:goals-for s))))))

(deftest league-standings
  (testing "Given a season's matches, When computing standings, Then order/points are exact"
    (let [table (q/standings ms "Brasileirão Série A" 2019)
          champ (first table)]
      (is (= 3 (count table)))
      (is (= "Flamengo" (:team champ)))
      (is (= 10 (:points champ)))
      (is (= 10 (:goal-diff champ)))
      (is (= ["Flamengo" "Palmeiras" "Santos"] (mapv :team table)))
      (is (= [10 5 1] (mapv :points table))))))

(deftest competition-statistics
  (testing "Given a competition/season, When aggregating, Then totals/rates match"
    (let [s (q/competition-stats ms {:competition "Brasileirão" :season 2019})]
      (is (= 6 (:matches s)))
      (is (= 15 (:total-goals s)))
      (is (= 2.5 (:avg-goals-per-match s)))
      (is (= 3 (:home-wins s)))
      (is (= 1 (:away-wins s)))
      (is (= 2 (:draws s))))))

(deftest biggest-wins-ranking
  (testing "Given matches, When ranking by margin, Then largest is first"
    (let [top (first (q/biggest-wins ms {:competition "Brasileirão" :season 2019}))]
      (is (= 5 (:margin top)))
      (is (= "Flamengo" (:home top)))
      (is (= "Santos" (:away top))))))

(deftest player-search
  (testing "Given players, When searching by club, Then club members return"
    (is (= 2 (count (q/search-players ps {:club "Flamengo"})))))

  (testing "When searching by nationality, Then sorted by rating desc"
    (let [br (q/search-players ps {:nationality "Brazil"})]
      (is (= 4 (count br)))
      (is (= "Neymar Jr" (:name (first br))))
      (is (= (reverse (sort (map :overall br))) (map :overall br)))))

  (testing "When filtering by min-overall, Then low-rated excluded"
    (is (= 1 (count (q/search-players ps {:nationality "Brazil" :min-overall 85})))))

  (testing "When taking top players, Then highest rated overall first"
    (is (= "Lionel Messi" (:name (first (q/top-players ps {})))))))

;; --- Smoke checks against the real datasets (cross-file, performance) -------

(deftest real-data-smoke
  (testing "Given real data, When querying a famous fixture, Then results return"
    (let [h2h (q/head-to-head "Flamengo" "Fluminense")]
      (is (pos? (:total (:record h2h)))))
    (let [neymar (q/search-players {:name "Neymar"})]
      (is (seq neymar))
      (is (every? #(= "Brazil" (:nationality %)) neymar)))
    (let [table (q/standings "Brasileirão Série A" 2019)
          champ (first table)]
      (is (>= (count table) 18))
      ;; 2019 Brasileirão was won by Flamengo with 90 pts in 38 games — a
      ;; published, verifiable result; confirms standings are not double-counted.
      (is (data/team-matches? "Flamengo" (:team champ)))
      (is (= 90 (:points champ)))
      (is (= 38 (:played champ)))))

  (testing "Given real data, When running an aggregate query, Then it is fast (<5s)"
    (let [start (System/nanoTime)
          _ (q/competition-stats {:competition "Brasileirão"})
          elapsed-ms (/ (- (System/nanoTime) start) 1e6)]
      (is (< elapsed-ms 5000) (str "elapsed-ms=" elapsed-ms)))))

File: experiments/adrianco/experiment-5/runs/language=go_model=claude-opus-4-8_tooling=none/rep1/internal/soccer/queries_test.go
// Context:
//   - BDD-style (Given/When/Then) tests for the loader and the query layer,
//     exercised against the real bundled Kaggle CSV files. They cover the five
//     capability areas in the specification: loading, match search, head-to-head,
//     team statistics, computed standings, league aggregates and player search.
//   - The 2019 Brasileirão Série A is used as a gold-standard fixture because its
//     final table is publicly known (Flamengo champion on 90 points with a
//     28-6-4 record), which lets us assert exact, meaningful numbers.
//   - The store is loaded once and shared across tests for speed.
package soccer

import (
	"path/filepath"
	"strings"
	"sync"
	"testing"
)

const testDataDir = "../../data/kaggle"

var (
	sharedStore *Store
	loadOnce    sync.Once
	loadErr     error
)

// store loads (once) and returns the shared store, skipping the test if the
// data files are not present.
func store(t *testing.T) *Store {
	t.Helper()
	loadOnce.Do(func() {
		sharedStore, loadErr = Load(testDataDir)
	})
	if loadErr != nil {
		t.Fatalf("loading data: %v", loadErr)
	}
	if sharedStore == nil || len(sharedStore.Matches) == 0 {
		t.Skipf("no data available at %s", filepath.Clean(testDataDir))
	}
	return sharedStore
}

func TestLoad_AllSourcesAndPlayers(t *testing.T) {
	// Given the bundled datasets
	s := store(t)

	// When loaded
	// Then a substantial number of matches and the full player set are present
	if len(s.Matches) < 10000 {
		t.Errorf("expected >10000 matches, got %d", len(s.Matches))
	}
	if len(s.Players) != 18207 {
		t.Errorf("expected 18207 players, got %d", len(s.Players))
	}

	// And all five canonical competitions are represented
	want := map[string]bool{
		CompSerieA: false, CompSerieB: false, CompSerieC: false,
		CompCopaDoBrasil: false, CompLibertadores: false,
	}
	for _, c := range s.Competitions() {
		if _, ok := want[c]; ok {
			want[c] = true
		}
	}
	for c, seen := range want {
		if !seen {
			t.Errorf("competition %q missing from loaded data", c)
		}
	}
}

func TestSearchMatches_BetweenTwoTeams(t *testing.T) {
	// Given the loaded match data
	s := store(t)

	// When searching for matches between Flamengo and Fluminense
	matches := s.SearchMatches(MatchQuery{Team: "Flamengo", Opponent: "Fluminense"})

	// Then a non-empty list is returned
	if len(matches) == 0 {
		t.Fatal("expected Fla-Flu matches, got none")
	}
	// And every match actually involves both teams
	for _, m := range matches {
		flHome := TeamMatches("Flamengo", m.HomeNorm)
		flAway := TeamMatches("Flamengo", m.AwayNorm)
		fluHome := TeamMatches("Fluminense", m.HomeNorm)
		fluAway := TeamMatches("Fluminense", m.AwayNorm)
		if !((flHome && fluAway) || (flAway && fluHome)) {
			t.Errorf("match does not involve both teams: %s", m.Line())
		}
	}
}

func TestSearchMatches_FilterBySeasonAndVenue(t *testing.T) {
	// Given the loaded match data
	s := store(t)

	// When searching Palmeiras home matches in the 2019 Série A
	matches := s.SearchMatches(MatchQuery{
		Team: "Palmeiras", Season: 2019, Competition: CompSerieA, Venue: "home",
	})

	// Then there are results and each is a 2019 Série A home game for Palmeiras
	if len(matches) == 0 {
		t.Fatal("expected Palmeiras 2019 home matches")
	}
	for _, m := range matches {
		if m.Season != 2019 || m.Competition != CompSerieA {
			t.Errorf("unexpected match: %s", m.Line())
		}
		if !TeamMatches("Palmeiras", m.HomeNorm) {
			t.Errorf("expected Palmeiras at home: %s", m.Line())
		}
	}
}

func TestHeadToHead_ConsistentRecord(t *testing.T) {
	// Given the loaded match data
	s := store(t)

	// When computing the Flamengo vs Fluminense head-to-head
	h := s.HeadToHead("Flamengo", "Fluminense")

	// Then there are matches and the W/D/L tallies are internally consistent
	if len(h.Matches) == 0 {
		t.Fatal("expected head-to-head matches")
	}
	played := 0
	for _, m := range h.Matches {
		if m.HasScore {
			played++
		}
	}
	if h.Team1Wins+h.Team2Wins+h.Draws != played {
		t.Errorf("W+D+L (%d) != played matches (%d)",
			h.Team1Wins+h.Team2Wins+h.Draws, played)
	}
}

func TestTeamStats_PointsAndWinRate(t *testing.T) {
	// Given the loaded match data
	s := store(t)

	// When requesting Flamengo's 2019 Série A record
	rec := s.TeamStats("Flamengo", 2019, CompSerieA, "")

	// Then it equals the publicly known champion record: 38 games, 28-6-4, 90 pts
	if rec.Played != 38 {
		t.Errorf("expected 38 matches, got %d", rec.Played)
	}
	if rec.Wins != 28 || rec.Draws != 6 || rec.Losses != 4 {
		t.Errorf("expected 28-6-4, got %d-%d-%d", rec.Wins, rec.Draws, rec.Losses)
	}
	if rec.Points() != 90 {
		t.Errorf("expected 90 points, got %d", rec.Points())
	}
	if rec.WinRate() <= 0.7 {
		t.Errorf("expected win rate >70%%, got %.1f%%", rec.WinRate()*100)
	}
}

func TestStandings_2019SerieAChampion(t *testing.T) {
	// Given the loaded match data
	s := store(t)

	// When computing the 2019 Série A table
	table := s.Standings(CompSerieA, 2019)

	// Then there are 20 teams and Flamengo is champion on 90 points
	if len(table) != 20 {
		t.Fatalf("expected 20 teams, got %d", len(table))
	}
	champ := table[0]
	if champ.Team != "Flamengo" {
		t.Errorf("expected champion Flamengo, got %q", champ.Team)
	}
	if champ.Points() != 90 {
		t.Errorf("expected champion on 90 points, got %d", champ.Points())
	}
	// And the table is sorted by points (non-increasing)
	for i := 1; i < len(table); i++ {
		if table[i-1].Points() < table[i].Points() {
			t.Errorf("table not sorted by points at row %d", i)
		}
	}
}

func TestLeagueAggregate_SaneAverages(t *testing.T) {
	// Given the loaded match data
	s := store(t)

	// When aggregating the 2019 Série A
	stats := s.LeagueAggregate(CompSerieA, 2019, 5)

	// Then it covers a full season and reports plausible figures
	if stats.Matches != 380 {
		t.Errorf("expected 380 matches in 2019 Série A, got %d", stats.Matches)
	}
	if stats.AvgGoals() < 1.5 || stats.AvgGoals() > 4.0 {
		t.Errorf("implausible average goals per match: %.2f", stats.AvgGoals())
	}
	if stats.HomeWinRate() <= 0 || stats.HomeWinRate() >= 1 {
		t.Errorf("implausible home win rate: %.2f", stats.HomeWinRate())
	}
	// And the biggest wins are sorted by goal margin
	for i := 1; i < len(stats.BiggestWins); i++ {
		prev := abs(stats.BiggestWins[i-1].HomeGoals - stats.BiggestWins[i-1].AwayGoals)
		cur := abs(stats.BiggestWins[i].HomeGoals - stats.BiggestWins[i].AwayGoals)
		if prev < cur {
			t.Errorf("biggest wins not sorted by margin at %d", i)
		}
	}
}

func TestSearchPlayers_TopBrazilian(t *testing.T) {
	// Given the loaded player data
	s := store(t)

	// When searching Brazilian players sorted by rating
	players := s.SearchPlayers(PlayerQuery{Nationality: "Brazil", Limit: 5})

	// Then there are results led by the highest-rated Brazilian, Neymar
	if len(players) == 0 {
		t.Fatal("expected Brazilian players")
	}
	if players[0].Name != "Neymar Jr" {
		t.Errorf("expected top Brazilian to be Neymar Jr, got %q", players[0].Name)
	}
	// And the list is sorted by overall rating (non-increasing)
	for i := 1; i < len(players); i++ {
		if players[i-1].Overall < players[i].Overall {
			t.Errorf("players not sorted by rating at %d", i)
		}
	}
}

func TestSearchPlayers_ByClubAndPosition(t *testing.T) {
	// Given the loaded player data
	s := store(t)

	// When filtering players at a Brazilian club present in the FIFA data
	// (the dataset is FIFA-19 era, which licenses Grêmio but not Flamengo).
	players := s.SearchPlayers(PlayerQuery{Club: "Grêmio", Limit: 50})

	// Then results exist and all belong to that club
	if len(players) == 0 {
		t.Fatal("expected players at Grêmio")
	}
	for _, p := range players {
		if !strings.Contains(normalizeText(p.Club), "gremio") {
			t.Errorf("player %q not at Grêmio (club=%q)", p.Name, p.Club)
		}
	}
}
